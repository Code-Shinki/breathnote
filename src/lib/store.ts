/* ＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/

  LIB -> STORE

  - microCMS から非同期処理で取得したデータを保存

  - CONFIG : サイト設定 (オブジェクト)
  - ALL_POSTS : 記事一覧 (配列)
  - ALL_CATEGORIES : カテゴリ一覧 (配列)
  - ALL_TAGS : タグ一覧 (配列)

＿/ ＿/ ＿/ ＿/ ＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/ */

import { AllCategoriesType, AllConfigType, AllPostsType, AllTagsType } from 'lib/types'

export const CONFIG = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllConfigType = await fetch('https://breathnote.microcms.io/api/v1/settings/config', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData
})()

export const ALL_POSTS = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllPostsType = await fetch('https://breathnote.microcms.io/api/v1/posts', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData.contents
})()

export const ALL_CATEGORIES = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllCategoriesType = await fetch('https://breathnote.microcms.io/api/v1/categories', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData.contents
})()

export const ALL_TAGS = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllTagsType = await fetch('https://breathnote.microcms.io/api/v1/tags', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData.contents
})()
