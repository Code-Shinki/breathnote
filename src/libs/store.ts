import axios from 'axios'
import { AllKeywords } from 'types/keyword'
import { AllPosts } from 'types/post'
import { API_ENDPOINT, API_KEY } from 'utils/env'

const KEY = { headers: { 'X-API-KEY': API_KEY } }

const fetchAllPosts = async () => {
  console.log('ALL_POSTS fire')

  // 一度に受け取る記事数
  const limit = 10

  // 総記事数等を含む最初の10記事を取得
  const posts: AllPosts = await axios
    .get(`${API_ENDPOINT}/posts?limit=${limit}`, KEY)
    .then((res) => res.data)
    .catch((err) => {
      throw new Error(`FETCH FAILED (fetchAllPosts() - First reqest) : ${err}`)
    })

  // APIを叩く回数
  const requestsCount = Math.floor(posts.totalCount / limit)

  // すべての記事を取得してposts.contentsに追加
  for (let i = 1; i <= requestsCount; i++) {
    const fetchData = await axios
      .get(`${API_ENDPOINT}/posts?offset=${limit * i}&limit=${limit}`, KEY)
      .then((res) => res.data)
      .catch((err) => {
        throw new Error(`FETCH FAILED (fetchAllPosts() - Roop ${i} reqest) : ${err}`)
      })
    posts.contents = posts.contents.concat(fetchData.contents)
  }

  // 記事を降順 (公開日が新しい順) に並び替え
  if (posts.contents.length > 1) {
    posts.contents.sort((a, b) => {
      if (a.publishedAt < b.publishedAt) {
        return 1
      } else {
        return -1
      }
    })
  }

  return posts
}

const fetchAllKeywords = async () => {
  console.log('ALL_KEYWORDS fire')

  // 一度に受け取るキーワード数
  const limit = 10

  // 総キーワード数等を含む最初の10データを取得
  const keywords: AllKeywords = await axios
    .get(`${API_ENDPOINT}/keywords?limit=${limit}`, KEY)
    .then((res) => res.data)
    .catch((err) => {
      throw new Error(`FETCH FAILED (fetchAllKeywords() - First reqest) : ${err}`)
    })

  // APIを叩く回数
  const requestsCount = Math.floor(keywords.totalCount / limit)

  // すべてのキーワードを取得してposts.contentsに追加
  for (let i = 1; i <= requestsCount; i++) {
    const fetchData = await axios
      .get(`${API_ENDPOINT}/keywords?offset=${limit * i}&limit=${limit}`, KEY)
      .then((res) => res.data)
      .catch((err) => {
        throw new Error(`FETCH FAILED (fetchAllKeywords() - Roop ${i} reqest) : ${err}`)
      })
    keywords.contents = keywords.contents.concat(fetchData.contents)
  }

  // キーワードを昇順 (アルファベット順) に並び替え
  if (keywords.contents.length > 1) {
    keywords.contents.sort((a, b) => {
      if (a.name > b.name) {
        return 1
      } else {
        return -1
      }
    })
  }

  return keywords
}

export const ALL_POSTS = fetchAllPosts()
export const ALL_KEYWORDS = fetchAllKeywords()
