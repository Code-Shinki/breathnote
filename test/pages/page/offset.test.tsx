import { cleanup, render, screen } from '@testing-library/react'
import React from 'react'
import Home from '../../../src/pages/page/[offset]'
import { ALL_POSTS } from '../../../src/scripts/store'

describe(`TOP PAGE 4/7 (pages/page/[offset].tsx)`, () => {
  afterEach(() => {
    cleanup
  })

  test('snapshot', async () => {
    const offset = 4
    const posts = (await ALL_POSTS).contents.slice(-3)
    const allPostLength = 42
    const option = {
      pageType: 'home',
      fullPath: `https://blog.shinki.net/page/${offset}`,
    }
    const { asFragment } = render(<Home posts={posts} allPostLength={allPostLength} option={option} offset={offset} />)
    const tree = asFragment()
    expect(tree).toMatchSnapshot()
  })

  test('integration', async () => {
    const offset = 4
    const posts = (await ALL_POSTS).contents.slice(-3)
    const allPostLength = 42
    const option = {
      pageType: 'home',
      fullPath: `https://blog.shinki.net/page/${offset}`,
    }
    render(<Home posts={posts} allPostLength={allPostLength} option={option} offset={offset} />)

    // ヘッダーが存在する
    expect(screen.getByRole('banner')).toBeInTheDocument()
    // メインコンテンツが存在する
    expect(screen.getByRole('main')).toBeInTheDocument()
    // サイドコンテンツ (SHARE, CONTENTS, MORE) が存在する
    expect(screen.getAllByRole('complementary')).toHaveLength(3)
    // フッターが存在する
    expect(screen.getByRole('contentinfo')).toBeInTheDocument()

    // h1属性が1個だけ存在する
    expect(screen.getByRole('heading', { level: 1 })).toBeInTheDocument()
    // 記事が1個以上存在する
    expect(screen.getAllByRole('article')).not.toHaveLength(0)
    // ページネーションが存在する
    expect(screen.getByRole('navigation', { name: 'pagination' })).toBeInTheDocument()
  })
})
