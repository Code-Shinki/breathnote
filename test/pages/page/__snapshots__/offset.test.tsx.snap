// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOP PAGE 4/7 (pages/page/[offset].tsx) snapshot 1`] = `
<DocumentFragment>
  <section
    class="root"
  >
    <article
      class="article"
    >
      <div
        class="thumbnail"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iODUwIiBoZWlnaHQ9IjQ0NSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="Material-UIとreact-springの併用で「Cannot update a component」"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </div>
      <time
        class="date"
        datetime="2021-03-19T14:47:17.358Z"
        itemprop="datepublished"
      >
        2021 / 03 / 19
      </time>
      <h2
        class="title"
      >
        <a
          href="/posts/react-spring-modal-state-error"
        >
          Material-UIとreact-springの併用で「Cannot update a component」
        </a>
      </h2>
      <ul
        class="keywords"
      >
        <li>
          React
        </li>
        <li>
          Material-UI
        </li>
      </ul>
    </article>
    <article
      class="article"
    >
      <div
        class="thumbnail"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iODUwIiBoZWlnaHQ9IjQ0NSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="Reactではステートの変更が即時反映されない場合がある"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </div>
      <time
        class="date"
        datetime="2021-03-16T16:08:28.810Z"
        itemprop="datepublished"
      >
        2021 / 03 / 17
      </time>
      <h2
        class="title"
      >
        <a
          href="/posts/react-state-not-reflected"
        >
          Reactではステートの変更が即時反映されない場合がある
        </a>
      </h2>
      <ul
        class="keywords"
      >
        <li>
          React
        </li>
      </ul>
    </article>
    <article
      class="article"
    >
      <div
        class="thumbnail"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iODUwIiBoZWlnaHQ9IjQ0NSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="環境変数を含むNext.jsのコードをJestでテストする"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </div>
      <time
        class="date"
        datetime="2021-03-05T06:00:02.829Z"
        itemprop="datepublished"
      >
        2021 / 03 / 05
      </time>
      <h2
        class="title"
      >
        <a
          href="/posts/nextjs-jest-environment-variable"
        >
          環境変数を含むNext.jsのコードをJestでテストする
        </a>
      </h2>
      <ul
        class="keywords"
      >
        <li>
          Next.js
        </li>
        <li>
          Jest
        </li>
      </ul>
    </article>
    <article
      class="article"
    >
      <div
        class="thumbnail"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iODUwIiBoZWlnaHQ9IjQ0NSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="ヘッドレスCMSから受け取ったマークダウン形式の記事でnext/imageを利用する"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </div>
      <time
        class="date"
        datetime="2021-02-27T06:00:02.752Z"
        itemprop="datepublished"
      >
        2021 / 02 / 27
      </time>
      <h2
        class="title"
      >
        <a
          href="/posts/nextjs-markdown-to-jsx"
        >
          ヘッドレスCMSから受け取ったマークダウン形式の記事でnext/imageを利用する
        </a>
      </h2>
      <ul
        class="keywords"
      >
        <li>
          Next.js
        </li>
        <li>
          Jamstack
        </li>
      </ul>
    </article>
    <article
      class="article"
    >
      <div
        class="thumbnail"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iODUwIiBoZWlnaHQ9IjQ0NSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="JamstackなNext.jsブログにプレビューモードを実装する"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </div>
      <time
        class="date"
        datetime="2021-02-23T14:16:00.797Z"
        itemprop="datepublished"
      >
        2021 / 02 / 23
      </time>
      <h2
        class="title"
      >
        <a
          href="/posts/nextjs-preview-mode"
        >
          JamstackなNext.jsブログにプレビューモードを実装する
        </a>
      </h2>
      <ul
        class="keywords"
      >
        <li>
          Next.js
        </li>
        <li>
          Jamstack
        </li>
      </ul>
    </article>
  </section>
  <nav
    aria-label="pagination"
    class="nav"
  >
    <ul
      class="ul"
    >
      <li
        class="li false"
      >
        <a
          class="link"
          href="/"
        >
          1
        </a>
      </li>
      <li
        class="dotted"
      >
        ...
      </li>
      <li
        class="li false"
      >
        <a
          class="link"
          href="/page/3"
        >
          3
        </a>
      </li>
      <li
        class="li li-current"
      >
        <a
          class="link"
          href="/page/4"
        >
          4
        </a>
      </li>
      <li
        class="li false"
      >
        <a
          class="link"
          href="/page/5"
        >
          5
        </a>
      </li>
      <li
        class="dotted"
      >
        ...
      </li>
      <li
        class="li false"
      >
        <a
          class="link"
          href="/page/7"
        >
          7
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
